src/lsp/server.ml

```diff
@@ -10,10 +10,12 @@ class dolmen_lsp_server =
     (* one env per document *)
     val buffers: (Lsp.Types.DocumentUri.t, State.t) Hashtbl.t = Hashtbl.create 32
 
+    val mutable prelude = []
```

Comment:
This should be set for each file rather than globally (so, it should be bound in the hashtable just above, which maps files to states).

Even if you want to use some preludes for a whole project/directory, it's probably simpler to make it so that your editor is configured to send the correct settings to the lsp for each file since the lsp server cannot know the exact situation.

Reply:
Not sure how to do that since the communication is only done through a defined set of notifications. The client can notify the server that a file was opened, closed or changed. It can also notify it that the settings have changed (regardless of whether or not there are files that are open or not) it just states that the settings have changed (for the workspace or globally) and it sends the new ones.

For the client to be able to send (possibly different) settings for each file then there needs to be file specific settings from the client side (which would be the same as using `include` directives).

I don't think that this is the intended use-case for the preludes, if you define a custom theory that was not yet added to the SMT-LIB or custom builtins and you want the server to process files that use them, you will probably have multiple files and use the same preludes for an entire workspace or at least whole repositories of those files. I don't think you'll have different preludes by file, in that case just use `include` directives.

The current version uses the same preludes for the whole workspace, alternatively the preludes can be applied to specific repositories then when the server receives a file, it can check whether it's in one of those repositories or not before determining whether or not to use the preludes.

Reply:
oh good point... I missed the fact that the settings sent by the editor were global and not tied to a file URI, :/

In that case, having a mutable field for the server seems reasonable.